src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,368 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class NullableWalker
+    {
+        /// <summary>
+        /// An immutable copy of Variables.
+        /// </summary>
+        [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+        internal sealed class VariablesSnapshot
+        {
+            internal readonly int Id;
+            internal readonly VariablesSnapshot? Container;
+            internal readonly Symbol? Symbol;
+            internal readonly ImmutableArray<KeyValuePair<VariableIdentifier, int>> VariableSlot;
+            internal readonly ImmutableDictionary<Symbol, TypeWithAnnotations> VariableTypes;
+
+            internal VariablesSnapshot(int id, VariablesSnapshot? container, Symbol? symbol, ImmutableArray<KeyValuePair<VariableIdentifier, int>> variableSlot, ImmutableDictionary<Symbol, TypeWithAnnotations> variableTypes)
+            {
+                Id = id;
+                Container = container;
+                Symbol = symbol;
+                VariableSlot = variableSlot;
+                VariableTypes = variableTypes;
+            }
+
+            internal bool TryGetType(Symbol symbol, out TypeWithAnnotations type)
+            {
+                return VariableTypes.TryGetValue(symbol, out type);
+            }
+
+            private string GetDebuggerDisplay()
+            {
+                var symbol = (object?)Symbol ?? "<null>";
+                return $"Id={Id}, Symbol={symbol}, Count={VariableSlot.Length}";
+            }
+        }
+
+        [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+        internal sealed class Variables
+        {
+            // Members of variables are tracked up to a fixed depth, to avoid cycles. The
+            // MaxSlotDepth value is arbitrary but large enough to allow most scenarios.
+            private const int MaxSlotDepth = 5;
+
+            private const int IdOffset = 16;
```

Comment:
>int [](start = 26, length = 3)

Should these be uints?

Reply:
> Should these be uints?

It shouldn't matter since the values are within the range of `int` and using `int` avoids casting.

---
In reply to: [560568771](https://github.com/dotnet/roslyn/pull/50417#discussion_r560568771) [](ancestors = 560568771)